indexer/crawler.go

```diff
@@ -64,114 +64,134 @@ type crawlWork struct {
-				act:        act,
-				initScrape: true,
-			}
-			c.todo[act.Uid] = cw
-			c.maplk.Unlock()
-
-			if next == nil {
-				next = cw
-				rs = c.repoSync
+			if nextDispatchedJob == nil {
+				nextDispatchedJob = crawlJob
+				dispatchQueue = c.repoSync
 			} else {
-				buffer = append(buffer, cw)
+				jobsAwaitingDispatch = append(jobsAwaitingDispatch, crawlJob)
 			}
-		case rs <- next:
-			c.maplk.Lock()
-			delete(c.todo, next.act.Uid)
-			c.inProgress[next.act.Uid] = next
-			c.maplk.Unlock()
+		case dispatchQueue <- nextDispatchedJob:
+			c.dequeueJob(nextDispatchedJob)
 
-			if len(buffer) > 0 {
-				next = buffer[0]
-				buffer = buffer[1:]
+			if len(jobsAwaitingDispatch) > 0 {
+				nextDispatchedJob = jobsAwaitingDispatch[0]
+				jobsAwaitingDispatch = jobsAwaitingDispatch[1:]
 			} else {
-				next = nil
-				rs = nil
+				nextDispatchedJob = nil
+				dispatchQueue = nil
 			}
-		case cw := <-c.catchup:
-			if next == nil {
-				next = cw
-				rs = c.repoSync
+		case catchupJob := <-c.catchup:
+			// CatchupJobs are for processing events that come in while a crawl is in progress
+			// They are lower priority than new crawls so we only add them to the queue if there isn't already a job in progress
+			if nextDispatchedJob == nil {
+				nextDispatchedJob = catchupJob
+				dispatchQueue = c.repoSync
 			} else {
-				buffer = append(buffer, cw)
+				jobsAwaitingDispatch = append(jobsAwaitingDispatch, catchupJob)
 			}
-
 		case uid := <-c.complete:
 			c.maplk.Lock()
+
 			job, ok := c.inProgress[uid]
 			if !ok {
 				panic("should not be possible to not have a job in progress we receive a completion signal for")
 			}
 			delete(c.inProgress, uid)
 
+			// If there are any subsequent jobs for this UID, add it back to the todo list or buffer.
+			// We're basically pumping the `next` queue into the `catchup` queue and will do this over and over until the `next` queue is empty.
 			if len(job.next) > 0 {
 				c.todo[uid] = job
 				job.initScrape = false
 				job.catchup = job.next
 				job.next = nil
-				if next == nil {
-					next = job
-					rs = c.repoSync
+				if nextDispatchedJob == nil {
+					nextDispatchedJob = job
+					dispatchQueue = c.repoSync
 				} else {
-					buffer = append(buffer, job)
+					jobsAwaitingDispatch = append(jobsAwaitingDispatch, job)
 				}
 			}
-
 			c.maplk.Unlock()
-
 		}
 	}
 }
 
+// enqueueJobForActor adds a new crawl job to the todo list if there isn't already a job in progress for this actor
+func (c *CrawlDispatcher) enqueueJobForActor(ai *models.ActorInfo) *crawlWork {
+	c.maplk.Lock()
+	defer c.maplk.Unlock()
+	_, ok := c.inProgress[ai.Uid]
+	if ok {
+		return nil
+	}
+
+	_, has := c.todo[ai.Uid]
+	if has {
+		return nil
+	}
+
+	crawlJob := &crawlWork{
+		act:        ai,
+		initScrape: true,
+	}
+	c.todo[ai.Uid] = crawlJob
+	return crawlJob
+}
+
+// dequeueJob removes a job from the todo list and adds it to the inProgress list
+func (c *CrawlDispatcher) dequeueJob(job *crawlWork) {
```

Comment:
i wouldnt neccesarily call this dequeue, its more 'beginJob' or something like that